File: Backend/EcoMonitor.Contracts/Models/PagedResult.cs
namespace EcoMonitor.Contracts.Models;

public sealed record PagedResult<T>
{
    public List<T> Items { get; private set; } = new();
    public int TotalCount { get; private set; }
    public int Page { get; private set; }
    public int PageSize { get; private set; }

    public int TotalPages => (int)Math.Ceiling(TotalCount / (double)PageSize);
    public bool HasNextPage => Page < TotalPages;
    public bool HasPreviousPage => Page > 1;

    public PagedResult(List<T> items, int totalCount, int page, int pageSize)
    {
        Items = items ?? new List<T>();
        TotalCount = totalCount;
        Page = page;
        PageSize = pageSize;
    }
}

File: FrontendFlutter/ecomonitor/lib/models/paged_result.dart
class PagedResult<T> {
  final List<T> items;
  final int totalCount;
  final int page;
  final int pageSize;

  PagedResult({
    required this.items,
    required this.totalCount,
    required this.page,
    required this.pageSize,
  });

  int get totalPages => (totalCount / pageSize).ceil();
  bool get hasNextPage => page < totalPages;
  bool get hasPrevious => page  > 1;

  factory PagedResult.fromJson(Map<String, dynamic> json, T Function(dynamic) fromJson) {
    var list = json['items'] as List;
    List<T> itemsList = list.map((i) => fromJson(i)).toList();
    return PagedResult(
      items: itemsList, 
      totalCount: json['totalCount'], 
      page: json['page'], 
      pageSize: json['pageSize'],
    );
  }
}

